### PR TITLE
CSP reporting

### LINKED ISSUE #729: Handle Content-Security Policy (CSP) violation reports
Content-Security Policy (CSP) is a new browser feature that makes it possible to restrict JavaScript files and other assets to a trusted set of origins:

http://www.html5rocks.com/en/tutorials/security/content-security-policy/
http://www.w3.org/TR/CSP/

CSP has a reporting mechanism whereby violations can be automatically POSTed by the browser to a reporting endpoint. It would be awesome if Sentry supported CSP violation reports, making it super easy for Sentry users to identify and track CSP violations (e.g. possible XSS attacks).

More on CSP's reporting mechanism:

http://www.html5rocks.com/en/tutorials/security/content-security-policy/#reporting
http://www.w3.org/TR/CSP/#sample-violation-report

### PR DESCRIPTION
**Pull request by mattrobenolt**
- [x] Code
- [x] Tests

Fixes GH-729

The only part I'm super not happy about is the `Http` interface, since technically, this shouldn't really have one, but it's being leverage to collect tags from the `User-Agent`. It just looks a bit misleading in the UI since we don't know the method, or any of the other headers, and replaying doesn't make sense.

One thing I'm happy about with this is the normalization with FF into Chrome (and everyone) else styles. FF sends reports a tiny bit differently than the rest of the world, and in my tests, I've gotten them to group together and act sensible.

This is leveraging a url like: `/api/<id>/csp-report/?sentry_key=fc1091a194da4309bdf4816628344602&sentry_version=5` which will have to be set in `report-uri` of their CSP header.

image

### CODE REVIEW

File: src/sentry/conf/server.py [hunk 1]
```diff
@@ -664,6 +665,7 @@ def create_partitioned_queues(name):
     'sentry.interfaces.Query': 'sentry.interfaces.query.Query',
     'sentry.interfaces.Http': 'sentry.interfaces.http.Http',
     'sentry.interfaces.User': 'sentry.interfaces.user.User',
+    'sentry.interfaces.CSP': 'sentry.interfaces.csp.CSP',
```
Comment (mattrobenolt) [inline src/sentry/conf/server.py]: Want this to be `Csp` instead of `CSP`?

Comment (mattrobenolt) [inline src/sentry/conf/server.py]: Nevermind, just did it. It fits the naming pattern of all the others, specifically like `Http`.

**Comment by dcramer** [comment]
Ping me in the morning to review but i agree on HTTP. IIRC JS tracebacks have the same exact issue though? We could very easily make the user agent stuff work with CSP.

File: src/sentry/static/sentry/app/components/events/eventEntries.jsx [hunk 2]
```diff
@@ -31,7 +31,8 @@ var EventEntries = React.createClass({
     exception: require("./interfaces/exception"),
     request: require("./interfaces/request"),
     stacktrace: require("./interfaces/stacktrace"),
-    template: require("./interfaces/template")
+    template: require("./interfaces/template"),
+    csp: require("./interfaces/csp"),
```
Comment (dcramer) [inline src/sentry/static/sentry/app/components/events/eventEntries.jsx]: trailing comma is invalid // @benvinegar how do we get lint for this

Comment (benvinegar) [inline src/sentry/static/sentry/app/components/events/eventEntries.jsx]: It's not invalid. Only IE7 and below throw error on this.

Comment (mattrobenolt) [inline src/sentry/static/sentry/app/components/events/eventEntries.jsx]: Oh, right. Yeah, @benvinegar, we talked about this before. You have a preference of no trailing commas just as convention, but a linter to enforce would be nice for those of us who don't do that instinctively. :)

Comment (benvinegar) [inline src/sentry/static/sentry/app/components/events/eventEntries.jsx]: Yeah, I added the rule to raven-js. I can do the same here.

**Comment by dcramer** [comment]
Regarding tests, can you setup a few fixtures for them that will have each of the browsers behavior just to make sure we dont end up in a similar place to how raven-js/tracekit was (where there were no tests for all the browser behavior)

**Comment by mattrobenolt** [comment]
> IIRC JS tracebacks have the same exact issue though?

Yeah, this is fair. I don't like how it's done there either, since it's slightly misleading. :imp: 

> We could very easily make the user agent stuff work with CSP.

I could call the plugins directly to extract the pieces, but this is fine for now imo. I'd rather let the processing pipeline work for us, instead of trying to outsmart it.

**Comment by joaomilho** [comment]
Hey, is there any way to send also tags in this endpoint?

**Comment by arthurlogilab** [comment]
hi @mattrobenolt this feature is really great, started using it today. 

I can't find any reference to this feature in https://docs.getsentry.com/ is this deliberate, if not, should I add an issue ?

**Comment by arthurlogilab** [comment]
oh, and by the way, it seems that firefox does not report when using an "external" report-uri, filed a bug https://bugzilla.mozilla.org/show_bug.cgi?id=1259393 (just putting this here as an info for those trying to use report-uri)

**Comment by mattrobenolt** [comment]
It's intentionally not documented yet since it's a bit unclear to us yet just how CSP/HPKP reporting fits into the product. In the meantime, I have documentation here https://gist.github.com/mattrobenolt/fe01bb9e37773289ae38

Also, thanks for the Firefox issue. I was unaware. I thought that worked before, but maybe I'm crazy. :)

**Comment by mattrobenolt** [comment]
@joaomilho I missed your question, there is not. Since CSP reports don't go through a client, this becomes a bit more difficult and requires custom stuff on our side.

The overall idea is that maybe we support adding all this stuff into the querystring, so you could do something like:  `&sentry_tags=a,b,c&sentry_release=abc`

This way we can leverage release tracking, etc, on CSP reports (which in practice is actually pretty important). But this hasn't been fully planned out yet.

**Comment by arthurlogilab** [comment]
@mattrobenolt I'm getting a probable bug on CSP handling. It's a follow up of that firefox bug I started submitting to mozilla. In fact the data is sent but I get a 403. The only difference I spotted so far is that Chrome uses type 'text/html' while Firefox uses 'json'. Do you have any idea about this ?

**Comment by mattrobenolt** [comment]
iirc Firefox fails to send us data that we're requiring and what is in the spec, I believe it's the `effective-directive`. If we don't have this, we'll reject. We need this to provide better grouping.

Also, Chrome sends the report as `application/csp-report`. At least, it used to, and Firefox send the incorrect type, `application/json`, but I submit a patch to fix that which has been merged. Sentry accepted both of these though.

**Comment by connorshea** [comment]
@mattrobenolt we'd like to use CSP reporting for Sentry at GitLab, any chance this is closer to being usable?

**Comment by mattrobenolt** [comment]
@connorshea You can opt into using this with the Early Adopter flag on your organization settings. Things are pretty solid imo.

**Comment by connorshea** [comment]
@mattrobenolt any documentation? I can't find it so I'm not sure how to get started with CSP reporting in Sentry :)

**Comment by mattrobenolt** [comment]
You need to be running at least Sentry 8.4 to have that option, I believe.

**Pull request merged**